### PR TITLE
Refactor conversations authentication (get and destroy)

### PR DESCRIPTION
**Pull request by spolu**
## Description

- Adds `canAccessConversation` check to `getConversation` (this one is risky as defense in depth will keep an eye post deploy, main risk are activities that didn't have the right auth. User flows should be safe)
- Adds a `unsafeRequestAllGroups` argument to `Authenticator.internalAdminForWorkspace`
- Leverages it in `destroyConversation` modifying call sites

## Risk

- Restriction of `getConversation` is a bit risky but much desirable. Tested various obvious flows in dev but hard to predict all API or activity-based flows here. The risk is worth the added protection.

## Deploy Plan

- deploy `front`

### CODE REVIEW

File: front/lib/api/assistant/conversation.ts [hunk 1]
```diff
@@ -344,6 +344,10 @@ export async function getConversation(
     return new Err(new ConversationError("conversation_not_found"));
   }
 
+  if (!canAccessConversation(auth, conversation)) {
```
Comment (flvndvd) [inline front/lib/api/assistant/conversation.ts:347]: This was leaving at the endpoint level, before?

Comment (spolu) [inline front/lib/api/assistant/conversation.ts:347]: somewhat high impact change

Comment (flvndvd) [reply front/lib/api/assistant/conversation.ts:347]: Just want to make sure that all downstream code do have proper auth object.

**Review by flvndvd — APPROVED** [review]

File: front/lib/api/assistant/conversation.ts (hunk 1 above)
Comment (spolu) [reply front/lib/api/assistant/conversation.ts:347]: Hmmm I don't think this was properly rolled out to all endpoints. Most endpoints rely on `getConversationWitoutContent` which does check for that. But here are a number of other endpoints that rely on getConversation that don't perform that check properly.

Eg, looking at pages/api/v1/w/[wId]/assistant/conversations/[cId]/index.ts not sure what prevents me from looking at a conversation I'm not supposed to have access to

Comment (flvndvd) [reply front/lib/api/assistant/conversation.ts:347]: 🤔 Pretty sure this was covered. Probably in the endpoint to fetch the pages of messages.

**Comment by spolu** [comment]
> Just want to make sure that all downstream code do have proper auth object.

Many of them

File: front/lib/api/assistant/conversation.ts (hunk 1 above)
Comment (spolu) [reply front/lib/api/assistant/conversation.ts:347]: Prove it to me :)

Comment (flvndvd) [reply front/lib/api/assistant/conversation.ts:347]: There is a check on fetching the message here: https://github.com/dust-tt/dust/blob/aeceabc74a6666ee55c416b7cf212ce0b9c675a8/front/lib/api/assistant/messages.ts#L383.

Which ensures that the current `auth` object has access to the message. Which is definitely not ideal.

File: front/lib/auth.ts [hunk 2]
```diff
@@ -524,9 +524,11 @@ export class Authenticator {
     });
   }
 
-  /* As above, with role `admin` */
+  /* As above, with role `admin`. Use requestAllGroups with care as it gives access to all groups
+   * within the workpsace. */
   static async internalAdminForWorkspace(
-    workspaceId: string
+    workspaceId: string,
+    unsafeRequestAllGroups = false
```
[the remark this answers is not in the document]
Comment (flvndvd) [reply front/lib/auth.ts]: > Maybe I'll go for workspaceId: string, options?: { dangerouslyRequestAllGroups }

We can do:
```
{workspaceId: string, dangerouslyRequestAllGroups?: boolean }
```

File: front/lib/api/assistant/conversation.ts (hunk 1 above)
Comment (flvndvd) [reply front/lib/api/assistant/conversation.ts:347]: Yep 🙌

File: front/lib/auth.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (spolu) [reply front/lib/auth.ts]: hmmm changes the signature in a way that is inconsistent with the other builder associated method. Went with option 3 which has minimal impact and does enforce typing

**Pull request merged**